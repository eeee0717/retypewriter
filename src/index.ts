import type { Diff } from 'diff-match-patch'
import { diff_match_patch as DMP } from 'diff-match-patch'

interface InsertPatch {
    type: 'insert'
    from: number
    text: string
}

interface RemovalPatch {
    type: 'removal'
    from: number
    length: number
}

type Patch = InsertPatch | RemovalPatch

export function diff(a: string, b: string): Diff[] {
    const differ = new DMP()
    const delta = differ.diff_main(a, b)
    differ.diff_cleanupSemantic(delta)
    return delta
}

export function calculatePatch(diff: Diff[]): Patch[] {
    const patches: Patch[] = []
    let index = 0
    for (const change of diff) {
        if (change[0] === 0) {
            index += change[1].length
            continue
        }

        else if (change[0] === -1) {
            const length = change[1].length
            patches.push({
                type: 'removal',
                from: index + length,
                length
            })
        }

        else if (change[0] === 1) {
            patches.push({
                type: 'insert',
                from: index,
                text: change[1],
            })
            index += change[1].length
        }
    }
    return patches

}

export function applyPatches(input: string, patches: Patch[]) {
    let output = input
    for (const patch of patches) {
        if (patch.type === 'insert')
            output = output.slice(0, patch.from) + patch.text + output.slice(patch.from)
        else if (patch.type === 'removal')
            output = output.slice(0, patch.from - patch.length) + output.slice(patch.from)
    }
    return {
        output,
    }
}
