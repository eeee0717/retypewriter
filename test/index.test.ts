import { describe, expect, it } from 'vitest'
import { applyPatches, calculatePatch, diff } from '../src'

const input = `
import { describe, expect, it } from 'vitest'
import type { Diff } from 'diff-match-patch'

describe('should', () => {
  it('exported', () => {
    expect(1).toEqual(1)
  })
})
`

const output = `
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(1).toEqual(1)
    expect(2).toEqual(2)
  })
})
`

describe('should', () => {
  it('exported', () => {
    const delta = diff(input, output)
    expect(delta).toMatchSnapshot('delta')
    const patches = calculatePatch(delta)
    expect(patches).toMatchSnapshot('patches')
    const applied = applyPatches(input, patches)
    expect(applied.output).toMatchSnapshot('applied')
    expect(applied.output).toEqual(output)
  })
})
