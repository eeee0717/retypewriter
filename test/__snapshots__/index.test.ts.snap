// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should > exported > applied 1`] = `
"
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(1).toEqual(1)
    expect(2).toEqual(2)
  })
})
"
`;

exports[`should > exported > delta 1`] = `
[
  [
    0,
    "
import { describe, expect, it } from 'vitest'",
  ],
  [
    -1,
    "
import type { Diff } from 'diff-match-patch'",
  ],
  [
    0,
    "

describe('should', () => {
  it('",
  ],
  [
    -1,
    "exported",
  ],
  [
    1,
    "one",
  ],
  [
    0,
    "', () => {
    expect(1).toEqual(1)
",
  ],
  [
    1,
    "    expect(2).toEqual(2)
",
  ],
  [
    0,
    "  })
})
",
  ],
]
`;

exports[`should > exported > patches 1`] = `
[
  {
    "from": 91,
    "length": 45,
    "type": "removal",
  },
  {
    "from": 89,
    "length": 8,
    "type": "removal",
  },
  {
    "from": 81,
    "text": "one",
    "type": "insert",
  },
  {
    "from": 120,
    "text": "    expect(2).toEqual(2)
",
    "type": "insert",
  },
]
`;
